src/System.Composition.Hosting/src/System/Composition/Hosting/Core/CompositionOperation.cs

```diff
@@ -0,0 +1,121 @@
+    public sealed class CompositionOperation : IDisposable
+    {
+        private List<Action> _nonPrerequisiteActions;
+        private List<Action> _postCompositionActions;
+        private object _sharingLock;
+
+        // Construct using Run() method.
+        private CompositionOperation() { }
+
+        /// <summary>
+        /// Execute a new composition operation starting within the specified lifetime
+        /// context, for the specified activator.
+        /// </summary>
+        /// <param name="outermostLifetimeContext">Context in which to begin the operation (the operation can flow
+        /// to the parents of the context if requried).</param>
+        /// <param name="compositionRootActivator">Activator that will drive the operation.</param>
+        /// <returns>The composed object graph.</returns>
+        public static object Run(LifetimeContext outermostLifetimeContext, CompositeActivator compositionRootActivator)
+        {
+            Requires.ArgumentNotNull(outermostLifetimeContext, "outermostLifetimeContext");
+            Requires.ArgumentNotNull(compositionRootActivator, "compositionRootActivator");
+
+            using (var operation = new CompositionOperation())
+            {
+                var result = compositionRootActivator(outermostLifetimeContext, operation);
+                operation.Complete();
+                return result;
+            }
+        }
+
+        /// <summary>
+        /// Called during the activation process to specify an action that can run after all
+        /// prerequesite part dependencies have been satisfied.
+        /// </summary>
+        /// <param name="action">Action to run.</param>
+        public void AddNonPrerequisiteAction(Action action)
+        {
+            if (action == null)
+                throw new ArgumentNullException("action");
+
+            if (_nonPrerequisiteActions == null)
+                _nonPrerequisiteActions = new List<Action>();
+
+            _nonPrerequisiteActions.Add(action);
+        }
+
+        /// <summary>
+        /// Called during the activation process to specify an action that must run only after
+        /// all composition has completed. See OnImportsSatisfiedAttribute.
+        /// </summary>
+        /// <param name="action">Action to run.</param>
+        public void AddPostCompositionAction(Action action)
+        {
+            Requires.ArgumentNotNull(action, "action");
+
+            if (_postCompositionActions == null)
+                _postCompositionActions = new List<Action>();
+
+            _postCompositionActions.Add(action);
+        }
+
+        internal void EnterSharingLock(object sharingLock)
+        {
+            Assumes.NotNull(sharingLock, "Sharing lock is required");
+
+            if (_sharingLock == null)
+            {
+                _sharingLock = sharingLock;
+                Monitor.Enter(sharingLock);
```

Comment:
Apologies in advance if I head off down the garden path here, I could be misunderstanding your query.

Though the same instance will never be used concurrently, multiple `CompositionOperation` instances running on _different_ threads may be active on the same scope concurrently. The scope owns the `_sharingLock`.

The first `CompositionOperation` to take `_sharingLock` will hold it until that operation ends, after which (in `Dispose()`) it will relinquish the lock and another `CompositionOperation` will proceed if any are waiting.

What might be the crucial missing info:

During a composition operation, `EnterSharingLock()` is called before beginning construction of a shared part within  a scope, _irrespective of whether the lock is already held_ by that operation, hence the subsequent calls to `EnterSharingLock()` being no-ops at that point. The name of the method could change to better reflect this - e.g. `EnsureSharingLockIsHeld()` would be an improvement I think.